File: 2024/TypeScript/01.ts
import * as fs from "fs";
import * as path from "path";

function readInput(): string {
  const inputPath = path.join(__dirname, "../input/01.txt");
  return fs.readFileSync(inputPath, "utf8").trim();
}

function part1(input: string): number {
  const pairs = input
    .split("\r\n")
    .map((line) => line.split("   ").map(Number));
  const left = pairs.map((pair) => pair[0]).sort();
  const right = pairs.map((pair) => pair[1]).sort();

  return left.reduce(
    (sum, leftNum, i) => sum + Math.abs(leftNum - right[i]),
    0
  );
}

function part2(input: string): number {
  const pairs = input
    .split("\r\n")
    .map((line) => line.split("   ").map(Number));
  const left = pairs.map((pair) => pair[0]);
  const right = pairs.map((pair) => pair[1]);

  const rightFreq = right.reduce((freq, num) => {
    freq.set(num, (freq.get(num) || 0) + 1);
    return freq;
  }, new Map<number, number>());

  return left.reduce((sum, leftNum) => {
    return sum + leftNum * (rightFreq.get(leftNum) || 0);
  }, 0);
}

const input = readInput();

console.log("Part 1:", part1(input));
console.log("Part 2:", part2(input));

File: 2024/Rust/01.rs
mod day_01 {
    use std::collections::HashMap;
    use std::fs;
    use std::path::{Path, PathBuf};

    fn read_input() -> String {
        let input_path: PathBuf = Path::new(file!()).parent().unwrap().join("../input/01.txt");
        fs::read_to_string(input_path)
            .expect("Failed to read input file")
            .trim()
            .to_string()
    }

    fn part1(input: &str) -> i32 {
        let pairs: Vec<(i32, i32)> = input
            .lines()
            .map(|line| {
                let nums: Vec<i32> = line
                    .split_whitespace()
                    .map(|s| s.parse().unwrap())
                    .collect();
                (nums[0], nums[1])
            })
            .collect();

        let mut left: Vec<i32> = pairs.iter().map(|(l, _)| *l).collect();
        let mut right: Vec<i32> = pairs.iter().map(|(_, r)| *r).collect();

        left.sort();
        right.sort();

        left.iter()
            .zip(right.iter())
            .map(|(l, r)| (l - r).abs())
            .sum()
    }

    fn part2(input: &str) -> i32 {
        let pairs: Vec<(i32, i32)> = input
            .lines()
            .map(|line| {
                let nums: Vec<i32> = line
                    .split_whitespace()
                    .map(|s| s.parse().unwrap())
                    .collect();
                (nums[0], nums[1])
            })
            .collect();

        let left: Vec<i32> = pairs.iter().map(|(l, _)| *l).collect();
        let right: Vec<i32> = pairs.iter().map(|(_, r)| *r).collect();

        let mut right_freq: HashMap<i32, i32> = HashMap::new();
        for num in right {
            *right_freq.entry(num).or_insert(0) += 1;
        }

        left.iter()
            .map(|left_num| left_num * right_freq.get(left_num).unwrap_or(&0))
            .sum()
    }

    pub fn run() {
        let input: String = read_input();

        println!("Part 1: {}", part1(&input));
        println!("Part 2: {}", part2(&input));
    }
}

pub use day_01::run;
